src/System.Text.Json/src/System/Text/Json/Serialization/WriteStackFrame.cs

```diff
@@ -35,41 +44,102 @@ internal void Initialize(Type type, JsonSerializerOptions options)
             }
         }
 
-        internal void Reset()
+        public void WriteObjectOrArrayStart(ClassType classType, Utf8JsonWriter writer, bool writeNull = false)
+        {
+            if (JsonPropertyInfo?._escapedName != null)
+            {
+                WriteObjectOrArrayStart(classType, JsonPropertyInfo?._escapedName, writer, writeNull);
+            }
+            else if (KeyName != null)
+            {
+                byte[] pooledKey = null;
+                byte[] utf8Key = Encoding.UTF8.GetBytes(KeyName);
+                int length = JsonWriterHelper.GetMaxEscapedLength(utf8Key.Length, 0);
+
+                Span<byte> escapedKey = length <= JsonConstants.StackallocThreshold ?
+                    stackalloc byte[length] :
+                    (pooledKey = ArrayPool<byte>.Shared.Rent(length));
+
+                JsonWriterHelper.EscapeString(utf8Key, escapedKey, 0, out int written);
+                Span<byte> propertyName = escapedKey.Slice(0, written);
+
+                WriteObjectOrArrayStart(classType, propertyName, writer, writeNull);
+
+                if (pooledKey != null)
+                {
+                    ArrayPool<byte>.Shared.Return(pooledKey);
+                }
+            }
+            else
+            {
+                Debug.Assert(writeNull == false);
+
+                // Write start without a property name.
+                if (classType == ClassType.Object || classType == ClassType.Dictionary)
+                {
+                    writer.WriteStartObject();
+                    StartObjectWritten = true;
+                }
+                else
+                {
+                    Debug.Assert(classType == ClassType.Enumerable);
+                    writer.WriteStartArray();
+                }
+            }
+        }
+
+        private void WriteObjectOrArrayStart(ClassType classType, Span<byte> propertyName, Utf8JsonWriter writer, bool writeNull)
```

Comment:
Should this be `ReadOnlySpan<byte>`?

Reply:
Yes it can be. Fyi it is called from only one other private method just above here, with a newly created Span<>.